targets/XPlatCppSdk/source/code/source/playfab/OneDSEventPipeline.cpp

```diff
@@ -0,0 +1,167 @@
+
+namespace PlayFab
+{
+    OneDSEventPipeline::OneDSEventPipeline(std::shared_ptr<OneDSEventPipelineSettings> settings) :
+        PlayFabEventPipeline(settings)
+    {
+    }
+
+    std::shared_ptr<OneDSEventPipelineSettings> OneDSEventPipeline::GetSettings() const
+    {
+        return std::dynamic_pointer_cast<OneDSEventPipelineSettings>(PlayFabEventPipeline::GetSettings());
+    }
+
+    void OneDSEventPipeline::SendBatch(size_t& batchCounter)
+    {
+        // create a WriteTelemetryEvents API request to send the batch
+        EventsModels::WriteEventsRequest batchReq;
+        for (const auto& eventEmitRequest : this->batch)
+        {
+            const auto& playFabEmitRequest = std::dynamic_pointer_cast<const PlayFabEmitEventRequest>(eventEmitRequest);
+            batchReq.Events.push_back(playFabEmitRequest->event->eventContents);
+        }
+
+        // add batch to flight tracking map
+        void* customData = reinterpret_cast<void*>(batchCounter); // used to track batches across asynchronous OneDS Events API
+        this->batchesInFlight[customData] = std::move(this->batch);
+        batchCounter++;
+
+        this->batch.clear(); // batch vector will be reused
+        this->batch.reserve(this->GetSettings()->maximalNumberOfItemsInBatch);
+
+        // authenticate OneDS API client as needed
+        if (!oneDSEventsApi.GetIsOneDSAuthenticated())
+        {
+            if (!AuthenticateOneDSApi(customData))
+                return;
+        }
+
+        // call OneDS Events API to send the batch
+        oneDSEventsApi.WriteTelemetryEvents(
+            batchReq,
+            std::bind(&OneDSEventPipeline::WriteTelemetryEventsApiCallback, this, std::placeholders::_1, std::placeholders::_2),
+            std::bind(&OneDSEventPipeline::WriteTelemetryEventsApiErrorCallback, this, std::placeholders::_1, std::placeholders::_2),
+            customData);
+    }
+
+    bool OneDSEventPipeline::AuthenticateOneDSApi(void* customData)
+    {
+        // Get OneDS context
+        bool operationComplete = false;
+        bool isOneDSAuthenticated = false;
+        EventsModels::TelemetryIngestionConfigRequest configRequest;
+        PlayFabTelemetryEventsAPI::GetTelemetryIngestionConfig(configRequest,
+            [&](const PlayFab::EventsModels::TelemetryIngestionConfigResponse& result, void* relayedCustomData)
+            {
+                oneDSEventsApi.SetCredentials("o:" + result.TenantId, result.IngestionKey);
+                isOneDSAuthenticated = true;
+                operationComplete = true;
+            },
+            [&](const PlayFab::PlayFabError& error, void* relayedCustomData)
+            {
+                // failed to get OneDS context info (including credentials), relay error to user
+                WriteTelemetryEventsApiErrorCallback(error, customData);
+                isOneDSAuthenticated = false;
+                operationComplete = true;
+            });
```

Comment:
While I understand most of the nuances about why this particular set of code will work, I am worried about it from a maintenance and code-cleanliness standpoint.

Specifically: Having async lambdas that refer to stack variables is dangerous in almost any other example. It works here, but it's a dangerous thing to mimic.

Reply:
I fail to see the point of relayedCustomData in this context.

Reply:
Changed pattern to use shared pointers instead of automatic variables as an example of a better (more reliable) memory model 

---
In reply to: [241204682](https://github.com/PlayFab/SDKGenerator/pull/358#discussion_r241204682) [](ancestors = 241204682)

Reply:
It's a part of method signature that the callback must follow

---
In reply to: [241205630](https://github.com/PlayFab/SDKGenerator/pull/358#discussion_r241205630) [](ancestors = 241205630)